File: sdk/python/dynaep/lattice/attractor_index.py
# ===========================================================================
# dynaep.lattice.attractor_index - Attractor Index with LSH
# OPT-007: O(1) expected-time attractor matching with LRU eviction.
# TLA+ invariant MemoryDoesNotAffectDecision preserved.
# ===========================================================================

from __future__ import annotations
import time
from dataclasses import dataclass, field
from typing import Dict, Any, Optional

from .lsh_index import LSHIndex
from .feature_extractor import extract_features, cosine_similarity, FEATURE_DIMENSION

import array


@dataclass
class LedgerAttractor:
    id: str
    features: Dict[str, Any]
    verdict: str  # "accepted" | "rejected"
    inserted_at: float = 0.0
    last_matched_at: float = 0.0


@dataclass
class AttractorStats:
    size: int
    inserts: int
    matches: int
    misses: int
    evictions: int
    avg_candidates: float


class AttractorIndex:
    """Attractor index using LSH for O(1) expected matching."""

    def __init__(
        self,
        max_attractors: int = 2000,
        similarity_threshold: float = 0.95,
        index_type: str = "lsh",
        lsh_tables: int = 8,
        lsh_hash_dimension: int = 4,
    ) -> None:
        self._max = max_attractors
        self._threshold = similarity_threshold
        self._index_type = index_type

        if index_type == "lsh":
            self._lsh: Optional[LSHIndex[str]] = LSHIndex(lsh_tables, lsh_hash_dimension, FEATURE_DIMENSION)
        else:
            self._lsh = None

        self._attractors: Dict[str, LedgerAttractor] = {}
        self._feature_vectors: Dict[str, array.array] = {}
        self._access_order: Dict[str, int] = {}
        self._access_counter = 0

        self._stat_inserts = 0
        self._stat_matches = 0
        self._stat_misses = 0
        self._stat_evictions = 0
        self._total_candidates = 0
        self._total_queries = 0

    def insert(self, attractor: LedgerAttractor) -> None:
        if len(self._attractors) >= self._max:
            self._evict_lru()
        features = extract_features(attractor.features)
        self._attractors[attractor.id] = attractor
        self._feature_vectors[attractor.id] = features
        self._access_order[attractor.id] = self._access_counter
        self._access_counter += 1
        if self._lsh:
            self._lsh.insert(attractor.id, features, attractor.id)
        self._stat_inserts += 1

    def find_match(self, proposal: Dict[str, Any]) -> Optional[LedgerAttractor]:
        proposal_features = extract_features(proposal)
        self._total_queries += 1

        if self._lsh:
            candidates = self._lsh.query(proposal_features)
        else:
            candidates = list(self._attractors.keys())

        self._total_candidates += len(candidates)

        best: Optional[LedgerAttractor] = None
        best_sim = -1.0
        for cid in candidates:
            cf = self._feature_vectors.get(cid)
            if cf is None:
                continue
            sim = cosine_similarity(proposal_features, cf)
            if sim >= self._threshold and sim > best_sim:
                best_sim = sim
                best = self._attractors.get(cid)

        if best:
            best.last_matched_at = time.time()
            self._access_order[best.id] = self._access_counter
            self._access_counter += 1
            self._stat_matches += 1
        else:
            self._stat_misses += 1
        return best

    def remove(self, attractor_id: str) -> None:
        self._attractors.pop(attractor_id, None)
        self._feature_vectors.pop(attractor_id, None)
        self._access_order.pop(attractor_id, None)
        if self._lsh:
            self._lsh.remove(attractor_id)

    def size(self) -> int:
        return len(self._attractors)

    def stats(self) -> AttractorStats:
        return AttractorStats(
            size=len(self._attractors),
            inserts=self._stat_inserts,
            matches=self._stat_matches,
            misses=self._stat_misses,
            evictions=self._stat_evictions,
            avg_candidates=self._total_candidates / self._total_queries if self._total_queries else 0,
        )

    def _evict_lru(self) -> None:
        if not self._access_order:
            return
        oldest_id = min(self._access_order, key=self._access_order.get)  # type: ignore
        self.remove(oldest_id)
        self._stat_evictions += 1

File: sdk/typescript/src/lattice/AttractorIndex.ts
// ===========================================================================
// @dynaep/core - Attractor Index with LSH
// OPT-007: Replaces brute-force O(A) attractor matching with
// locality-sensitive hashing for O(1) expected lookup. Includes
// capacity management with LRU eviction.
//
// The TLA+ invariant MemoryDoesNotAffectDecision is preserved:
// LSH false negatives send proposals to cold-path (still correct),
// LSH false positives are eliminated by exact similarity verification.
// ===========================================================================

import { LSHIndex } from "./LSHIndex";
import { extractFeatures, cosineSimilarity, FEATURE_DIMENSION, type FeatureSource } from "./FeatureExtractor";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface AttractorConfig {
  maxAttractors: number;
  similarityThreshold: number;
  indexType: "lsh" | "brute_force";
  lshTables: number;
  lshHashDimension: number;
}

export interface LedgerAttractor {
  id: string;
  features: FeatureSource;
  verdict: "accepted" | "rejected";
  insertedAt: number;
  lastMatchedAt: number;
}

export interface AttractorStats {
  size: number;
  inserts: number;
  matches: number;
  misses: number;
  evictions: number;
  avgCandidates: number;
}

// ---------------------------------------------------------------------------
// AttractorIndex
// ---------------------------------------------------------------------------

export class AttractorIndex {
  private readonly config: AttractorConfig;
  private readonly lshIndex: LSHIndex<string> | null;
  private readonly attractors: Map<string, LedgerAttractor>;
  private readonly featureVectors: Map<string, Float32Array>;
  private readonly accessOrder: Map<string, number>;
  private accessCounter: number;

  // Stats tracking
  private statInserts: number;
  private statMatches: number;
  private statMisses: number;
  private statEvictions: number;
  private totalCandidates: number;
  private totalQueries: number;

  constructor(config: Partial<AttractorConfig> = {}) {
    this.config = {
      maxAttractors: config.maxAttractors ?? 2000,
      similarityThreshold: config.similarityThreshold ?? 0.95,
      indexType: config.indexType ?? "lsh",
      lshTables: config.lshTables ?? 8,
      lshHashDimension: config.lshHashDimension ?? 4,
    };

    this.attractors = new Map<string, LedgerAttractor>();
    this.featureVectors = new Map<string, Float32Array>();
    this.accessOrder = new Map<string, number>();
    this.accessCounter = 0;

    if (this.config.indexType === "lsh") {
      this.lshIndex = new LSHIndex<string>(
        this.config.lshTables,
        this.config.lshHashDimension,
        FEATURE_DIMENSION,
      );
    } else {
      this.lshIndex = null;
    }

    this.statInserts = 0;
    this.statMatches = 0;
    this.statMisses = 0;
    this.statEvictions = 0;
    this.totalCandidates = 0;
    this.totalQueries = 0;
  }

  /**
   * Insert an attractor into the index. If at capacity, evicts the
   * least recently used (matched) attractor first.
   */
  insert(attractor: LedgerAttractor): void {
    // Evict LRU if at capacity
    if (this.attractors.size >= this.config.maxAttractors) {
      this.evictLRU();
    }

    const features = extractFeatures(attractor.features);
    this.attractors.set(attractor.id, attractor);
    this.featureVectors.set(attractor.id, features);
    this.accessOrder.set(attractor.id, this.accessCounter++);

    if (this.lshIndex) {
      this.lshIndex.insert(attractor.id, features, attractor.id);
    }

    this.statInserts++;
  }

  /**
   * Find the best matching attractor for a proposal.
   * Uses LSH for candidate generation (O(1) per table) then
   * exact cosine similarity verification on candidates.
   *
   * Returns null if no attractor exceeds the similarity threshold.
   * LSH false negatives are acceptable (proposal goes to cold path).
   */
  findMatch(proposal: FeatureSource): LedgerAttractor | null {
    const proposalFeatures = extractFeatures(proposal);
    this.totalQueries++;

    let candidates: string[];

    if (this.lshIndex) {
      // LSH candidate generation
      candidates = this.lshIndex.query(proposalFeatures);
    } else {
      // Brute-force: all attractors are candidates
      candidates = Array.from(this.attractors.keys());
    }

    this.totalCandidates += candidates.length;

    let bestAttractor: LedgerAttractor | null = null;
    let bestSimilarity = -1;

    for (const candidateId of candidates) {
      const candidateFeatures = this.featureVectors.get(candidateId);
      if (!candidateFeatures) continue;

      const similarity = cosineSimilarity(proposalFeatures, candidateFeatures);
      if (similarity >= this.config.similarityThreshold && similarity > bestSimilarity) {
        bestSimilarity = similarity;
        bestAttractor = this.attractors.get(candidateId) ?? null;
      }
    }

    if (bestAttractor) {
      // Update LRU tracking
      bestAttractor.lastMatchedAt = Date.now();
      this.accessOrder.set(bestAttractor.id, this.accessCounter++);
      this.statMatches++;
    } else {
      this.statMisses++;
    }

    return bestAttractor;
  }

  /**
   * Remove a specific attractor from the index.
   */
  remove(attractorId: string): void {
    this.attractors.delete(attractorId);
    this.featureVectors.delete(attractorId);
    this.accessOrder.delete(attractorId);

    if (this.lshIndex) {
      this.lshIndex.remove(attractorId);
    }
  }

  /**
   * Return the number of attractors in the index.
   */
  size(): number {
    return this.attractors.size;
  }

  /**
   * Return index statistics.
   */
  stats(): AttractorStats {
    return {
      size: this.attractors.size,
      inserts: this.statInserts,
      matches: this.statMatches,
      misses: this.statMisses,
      evictions: this.statEvictions,
      avgCandidates: this.totalQueries > 0
        ? this.totalCandidates / this.totalQueries
        : 0,
    };
  }

  // -------------------------------------------------------------------------
  // Private
  // -------------------------------------------------------------------------

  /**
   * Evict the least recently used attractor (by last access time).
   */
  private evictLRU(): void {
    let oldestId: string | null = null;
    let oldestAccess = Infinity;

    for (const [id, access] of this.accessOrder) {
      if (access < oldestAccess) {
        oldestAccess = access;
        oldestId = id;
      }
    }

    if (oldestId !== null) {
      this.remove(oldestId);
      this.statEvictions++;
    }
  }
}
